vyper/ir/optimizer.py

```diff
@@ -1,191 +1,389 @@
+        # x + 0 == x - 0 == x | 0 == x ^ 0 == x
+        new_val = args[0].value
+        new_args = args[0].args
+
+    elif binop in {"sub", "xor"} and _conservative_eq(args[0], args[1]):
+        # x - x == x ^ x == 0
+        new_val = 0
+        new_args = []
+
+    # TODO associativity rules
+
+    elif binop in {"mul", "div", "sdiv", "mod", "smod", "and"} and _int(args[1]) == 0:
+        new_val = 0
+        new_args = []
+
+    elif binop in {"mod", "smod"} and _int(args[1]) == 1:
+        new_val = 0
+        new_args = []
+
+    elif binop in {"mul", "div", "sdiv"} and _int(args[1]) == 1:
+        new_val = args[0].value
+        new_args = args[0].args
+
+    # x * -1 == 0 - x
+    elif binop in {"mul", "sdiv"} and _int(args[1], SIGNED) == -1:
+        new_val = "sub"
+        new_args = [0, args[0]]
+
+    # maybe OK:
+    # elif binop == "div" and _int(args[1], UNSIGNED) == MAX_UINT256:
+    #    # (div x (2**256 - 1)) == (eq x (2**256 - 1))
+    #    new_val = "eq"
+    #    args = args
+
+    elif binop in {"mod", "div", "mul"} and _is_int(args[1]) and is_power_of_two(_int(args[1])):
+        assert unsigned == UNSIGNED, "something's not right."
+        # shave two gas off mod/div/mul for powers of two
+        if binop == "mod":
+            new_val = "and"
+            new_args = [args[0], _int(args[1]) - 1]
+        if binop == "div" and version_check(begin="constantinople"):
+            new_val = "shr"
+            # recall shr/shl have unintuitive arg order
+            new_args = [int_log2(_int(args[1])), args[0]]
+        # note: no rule for sdiv since it rounds differently from sar
+        if binop == "mul" and version_check(begin="constantinople"):
+            new_val = "shl"
+            new_args = [int_log2(_int(args[1])), args[0]]
+
+    elif binop in {"and", "or", "xor"} and _int(args[1]) == 2 ** 256 - 1:
+        if binop == "and":
+            new_val = args[0].value
+            new_args = args[0].args
+        elif binop == "xor":
+            new_val = "not"
+            new_args = [args[0]]
+        elif binop == "or":
+            new_val = args[1].value
+            new_args = []
+        else:  # pragma: nocover
+            raise CompilerPanic("unreachable")
+
+    ##
+    # COMPARISONS
+    ##
+
+    # note: in places where truthy is accepted, sequences of
+    # ISZERO ISZERO will be optimized out, so we try to rewrite
+    # some operations to include iszero
+    # (note ordering; truthy optimizations should come first
+    # to avoid getting clobbered by other branches)
+    # TODO: rethink structure of is_truthy optimizations.
+    elif is_truthy:
+        if binop == "eq":
+            # x == 0xff...ff => ~x == 0
+            if _int(args[1], UNSIGNED) == 2 ** 256 - 1:
+                new_val = "iszero"
+                new_args = [["not", args[0]]]
+            else:
+                # (eq x y) has the same truthyness as (iszero (sub x y))
```

Comment:
Overflow shouldn't affect this, right?

Reply:
no -- for instance, `sub 1 2` returns `-1`
![Screenshot from 2022-05-09 23-37-48](https://user-images.githubusercontent.com/3867501/167502896-909333d7-fc22-4ddc-a5d0-2a953542b018.png)